File: backend/app/schemas/document.py
from pydantic import BaseModel, Field
from enum import Enum
from typing import Optional
import uuid
from datetime import datetime


class DocumentType(str, Enum):
    RENTAL_AGREEMENT = "rental_agreement"
    GOVERNMENT_FORM = "government_form"
    HEALTHCARE = "healthcare"
    BANKING = "banking"
    EMPLOYMENT_CONTRACT = "employment_contract"
    IMMIGRATION = "immigration"
    EDUCATIONAL = "educational"
    UNKNOWN = "unknown"


class RiskLevel(str, Enum):
    LOW = "low"
    MEDIUM = "medium"
    HIGH = "high"


class GlossaryItem(BaseModel):
    term: str
    definition: str
    cultural_note: Optional[str] = None


class SimplifiedSection(BaseModel):
    original: str
    simplified: str
    section_title: Optional[str] = None


class Deadline(BaseModel):
    description: str
    date: Optional[str] = None
    urgency: RiskLevel = RiskLevel.MEDIUM


class Risk(BaseModel):
    description: str
    level: RiskLevel
    recommendation: Optional[str] = None


class ActionItem(BaseModel):
    action: str
    priority: RiskLevel
    due_by: Optional[str] = None


class DocumentAnalysis(BaseModel):
    document_type: DocumentType
    language_detected: str = "en"
    summary: str
    important_points: list[str] = Field(default_factory=list)
    deadlines: list[Deadline] = Field(default_factory=list)
    risks: list[Risk] = Field(default_factory=list)
    simplified_sections: list[SimplifiedSection] = Field(default_factory=list)
    glossary: list[GlossaryItem] = Field(default_factory=list)
    action_items: list[ActionItem] = Field(default_factory=list)
    full_translation: Optional[str] = None
    confidence_score: float = Field(default=0.85, ge=0.0, le=1.0)
    processing_time_ms: int = 0


class ProcessingStatus(str, Enum):
    PENDING = "pending"
    EXTRACTING_TEXT = "extracting_text"
    CLASSIFYING = "classifying"
    ANALYZING = "analyzing"
    COMPLETE = "complete"
    ERROR = "error"


class DocumentRecord(BaseModel):
    id: str = Field(default_factory=lambda: str(uuid.uuid4()))
    filename: str
    file_size: int
    mime_type: str
    target_language: str = "en"
    status: ProcessingStatus = ProcessingStatus.PENDING
    progress_pct: int = 0
    current_step: str = ""
    error: Optional[str] = None
    analysis: Optional[DocumentAnalysis] = None
    extracted_text: Optional[str] = None
    uploaded_at: datetime = Field(default_factory=datetime.utcnow)

File: frontend/src/types/document.ts
// Core document types that mirror the backend Pydantic schemas

export type DocumentType =
  | "rental_agreement"
  | "government_form"
  | "healthcare"
  | "banking"
  | "employment_contract"
  | "immigration"
  | "educational"
  | "unknown";

export type RiskLevel = "low" | "medium" | "high";

export interface GlossaryItem {
  term: string;
  definition: string;
  cultural_note?: string;
}

export interface SimplifiedSection {
  original: string;
  simplified: string;
  section_title?: string;
}

export interface Deadline {
  description: string;
  date?: string;
  urgency: RiskLevel;
}

export interface Risk {
  description: string;
  level: RiskLevel;
  recommendation?: string;
}

export interface ActionItem {
  action: string;
  priority: RiskLevel;
  due_by?: string;
}

export interface DocumentAnalysis {
  document_type: DocumentType;
  language_detected: string;
  summary: string;
  important_points: string[];
  deadlines: Deadline[];
  risks: Risk[];
  simplified_sections: SimplifiedSection[];
  glossary: GlossaryItem[];
  action_items: ActionItem[];
  full_translation?: string;
  confidence_score: number;
  processing_time_ms: number;
}

export interface UploadedDocument {
  id: string;
  filename: string;
  file_size: number;
  mime_type: string;
  uploaded_at: string;
  status: ProcessingStatus;
  analysis?: DocumentAnalysis;
}

export type ProcessingStatus =
  | "pending"
  | "extracting_text"
  | "classifying"
  | "analyzing"
  | "complete"
  | "error";

export interface ProcessingStep {
  id: string;
  label: string;
  description: string;
  status: "pending" | "active" | "complete" | "error";
}

export const PROCESSING_STEPS: ProcessingStep[] = [
  {
    id: "extract",
    label: "Reading Document",
    description: "Extracting text from your document...",
    status: "pending",
  },
  {
    id: "classify",
    label: "Identifying Document Type",
    description: "Figuring out what kind of document this is...",
    status: "pending",
  },
  {
    id: "simplify",
    label: "Simplifying Language",
    description: "Translating complex terms into plain language...",
    status: "pending",
  },
  {
    id: "analyze",
    label: "Finding Key Information",
    description: "Identifying deadlines, risks, and actions for you...",
    status: "pending",
  },
  {
    id: "complete",
    label: "Preparing Your Results",
    description: "Almost done! Getting everything ready...",
    status: "pending",
  },
];
